File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/TimestampedDataAttributeInfo.kt
package com.jetbrains.signatureverifier.serialization

import kotlinx.serialization.Serializable
import org.bouncycastle.asn1.DLSequence
import org.bouncycastle.asn1.cms.Attribute

// 1.3.6.1.4.1.311.10.3.28
@Serializable
data class TimestampedDataAttributeInfo(
  val identifier: TextualInfo,
  val content: List<TextualInfo>
): AttributeInfo {

  override fun toAttributeDLSequence(): DLSequence =
    listOf(
      identifier.toPrimitive(),
      content.map { it.toPrimitive() }.toDLSet()
    ).toDLSequence()

  constructor(attribute: Attribute): this(
    TextualInfo.getInstance(attribute.attrType),
    attribute.attributeValues.map { TextualInfo.getInstance(it) }
  )
}

File: net/JetBrains.Serialization/src/Dataholders/TimestampedDataAttributeInfo.cs
using Newtonsoft.Json;
using Org.BouncyCastle.Asn1;
using Attribute = Org.BouncyCastle.Asn1.Cms.Attribute;
using JetBrains.Serialization;
using System.Collections.Generic;
using System.Linq;

[JsonObject(MemberSerialization.Fields)]
public class TimestampedDataAttributeInfo : AttributeInfo
{
    public override TextualInfo Identifier { get; }
    public List<TextualInfo> Content { get; }

    public TimestampedDataAttributeInfo(Attribute attribute)
    {
        Identifier = TextualInfo.GetInstance(attribute.AttrType);
        Content = attribute.AttrValues.ToArray().Select(item => TextualInfo.GetInstance(item)).ToList();
    }

    public override Asn1Encodable GetPrimitiveContent() => Content.ToPrimitiveDerSet();
}